crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,295 @@
+                    match next_committee {
+                        None => {
+                            self.metrics.last_executed_checkpoint.set(seq_num as i64);
+                            self.checkpoint_store
+                                .update_highest_executed_checkpoint(&checkpoint)
+                                .unwrap();
+
+                            let highest_synced = self
+                                .checkpoint_store
+                                .get_highest_synced_checkpoint_seq_number().unwrap()
+                                .unwrap_or(0);
+                            // If we got here, `highest_scheduled_checkpoint` must have been set,
+                            // so safe to unwrap
+                            if self.highest_scheduled_checkpoint.unwrap() < highest_synced && !self.end_of_epoch {
+                                self.schedule_checkpoint(seq_num.saturating_add(1), &mut finished).await;
+                            }
+                        }
+                        // Last checkpoint of epoch -- Note that we must not update the
+                        // highest executed watermark until after reconfig has completed!
+                        Some(committee) => return Some((checkpoint, committee)),
+                    }
+                }
+                // Read from StateSync channel if we have capacity to schedule more checkpoints.
+                // If we're at end of epoch, skip this branch so that we can effectively await
+                // execution of all remaining checkpoints of the epoch (which should be scheduled)
+                // and call reconfig.
+                received = self.mailbox.recv(), if finished.len() < task_limit && !self.end_of_epoch => match received {
+                    Ok(checkpoint) => {
+                        let last_to_exec = checkpoint.sequence_number() as u64;
+                        let highest_executed_checkpoint = self.checkpoint_store.get_highest_executed_checkpoint_seq_number().unwrap_or_else(
+                            |err| {
+                                panic!("Failed to read highest executed checkpoint sequence number: {:?}", err)
+                            }
+                        );
+
+                        // Note that either of these can be higher. If the node crashes with many
+                        // scheduled tasks, then the in-memory watermark starts as None, but the
+                        // persistent watermark is set, hence we start there. If we get a new
+                        // messsage with checkpoints tasks scheduled, then the in-memory watermark
+                        // will be greater, and hence we start from there.
+                        let next_to_exec = std::cmp::max(
+                            highest_executed_checkpoint
+                                .map(|highest| highest.saturating_add(1))
+                                .unwrap_or(0),
```

Comment:
Do we always unwrap them to 0 when None? If so, why do we need to keep them as None instead of just 0 to begin with in memory?

Reply:
because we need to account for the case where we haven't executed any checkpoints. None is different than checkpoint 0.

Reply:
I searched around the use of both the local varaible highest_executed_checkpoint and self.highest_scheduled_checkpoint, I do not see how a None case is handled differently than 0.

Reply:
This exact line of code treats the None case differently than 0. If highest executed checkpoint is None, the next checkpoint we want to execute is 0. If highest executed checkpoint is 0, then the next checkpoint we want to execute is 1.

Reply:
I am assuming that `checkpoint_sequence_number` is 0 indexed. So `None` indicates we have not executed any, and `0` indicates we have executed checkpoint 0. I am setting to 0 here in the case of `None` because that would be the next to execute in this case, since this watermark is persistent and monotonically increasing (`None` must mean we have never executed checkpoint 0)